### PR TITLE
Fix sqlite issue 1580

### PR DESCRIPTION
**Pull request by trevorsibanda**
#1580

### CODE REVIEW

File: slick/src/main/scala/slick/jdbc/SQLiteProfile.scala [hunk 1]
```diff
@@ -251,6 +251,9 @@ trait SQLiteProfile extends JdbcProfile {
       override def valueToSQLLiteral(value: Time) = value.getTime.toString
     }
     class TimestampJdbcType extends super.TimestampJdbcType {
+      override def getValue(r: java.sql.ResultSet, idx: Int) = try r.getTimestamp(idx) catch {
+        case e: java.sql.SQLException => java.sql.Timestamp.valueOf(  r.getString(idx) )
+      }
```
Comment (cvogt) [inline slick/src/main/scala/slick/jdbc/SQLiteProfile.scala]: A test would be great, so if somebody breaks it, we'll notice. Is that doable?

Comment (trevorsibanda) [inline slick/src/main/scala/slick/jdbc/SQLiteProfile.scala]: @cvogt Added test

**Comment by cvogt** [comment]
Great stuff :). Would be great if we could add one more thing: Insert a row into this new table in https://github.com/slick/slick/blob/408f2e55aa0a7553a445c83b7ee5ef159f805694/slick-testkit/src/test/scala/slick/test/codegen/GeneratedCodeTest.scala#L84 and read the value back out and check if it is the one we had set as the default.

**Comment by typesafehub-validator** [comment]
Hi @invalid-email-address,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

File: slick/src/main/scala/slick/jdbc/SQLiteProfile.scala [hunk 2]
```diff
@@ -98,14 +98,16 @@ trait SQLiteProfile extends JdbcProfile {
     override def createColumnBuilder(tableBuilder: TableBuilder, meta: MColumn): ColumnBuilder = new ColumnBuilder(tableBuilder, meta) {
       /** Regex matcher to extract name and length out of a db type name with length ascription */
       final val TypePattern = "^([A-Z]+)(\\(([0-9]+)\\))?$".r
+      final val TimestampPattern = """^"(.*)"""".r
```
Comment (lastland) [inline slick/src/main/scala/slick/jdbc/SQLiteProfile.scala]: Is this sufficient to cover all time strings allowed by SQLite (See: https://www.sqlite.org/lang_datefunc.html section "Time Strings")? For example, SQLite allows the following time string "2456572.84952685", while `java.sql.Timestamp` expects a string in format "yyyy-[m]m-[d]d hh:mm:ss[.f...]".

File: slick-testkit/src/codegen/resources/dbs/sqlite.sql [hunk 3]
```diff
@@ -1,2 +1,3 @@
 CREATE TABLE "SUPPLIERS" ("SUP_ID" INTEGER NOT NULL,"SUP_NAME" VARCHAR(254) NOT NULL,"STREET" VARCHAR(254) NOT NULL,"CITY" VARCHAR(254) NOT NULL,"STATE" VARCHAR(254) NOT NULL,"ZIP" VARCHAR(254) NOT NULL);
 CREATE TABLE "COFFEES" ("COF_NAME" VARCHAR(254) NOT NULL,"SUP_ID" INTEGER NOT NULL,"PRICE" DOUBLE NOT NULL,"SALES" INTEGER NOT NULL,"TOTAL" INTEGER NOT NULL,constraint "SUP_FK" foreign key("SUP_ID") references "SUPPLIERS"("SUP_ID") on update NO ACTION on delete NO ACTION);
+CREATE TABLE "TIMESTAMP"("TS" TIMESTAMP NOT NULL DEFAULT "2016-07-29 09:14:56");
```
Comment (lastland) [inline slick-testkit/src/codegen/resources/dbs/sqlite.sql]: Does it make sense to add more tests on different time strings format allowed by SQLite?

Comment (trevorsibanda) [inline slick-testkit/src/codegen/resources/dbs/sqlite.sql]: @lastland Added more tests

File: slick/src/main/scala/slick/jdbc/SQLiteProfile.scala [hunk 4]
```diff
@@ -98,14 +98,24 @@ trait SQLiteProfile extends JdbcProfile {
...
       override def varying = dbType == Some("VARCHAR")
-      override def default = meta.columnDef.map((_,tpe)).collect{
+      override def default = meta.columnDef.map((_,tpe)).collect{ 
         case ("null",_)  => Some(None) // 3.7.15-M1
+        case (v , "java.sql.Timestamp") => Some(Some( v match{
+            case TimestampPattern(ts,ts_unix) => if( ts != null  ){
+                if(ts contains ":" )
+                  if( ts contains "-") java.sql.Timestamp.valueOf(ts) else new java.sql.Timestamp(  java.sql.Time.valueOf(ts).getTime )
+                else if( ts contains "-" )
+                  new java.sql.Timestamp(java.sql.Date.valueOf(ts).getTime)
```
Comment (lastland) [inline slick/src/main/scala/slick/jdbc/SQLiteProfile.scala]: How about the timestamp of a specific time zone? like `2013-10-07T08:23:19.120Z` and `2013-10-07 04:23:19.120-04:00` shown [here](https://www.sqlite.org/lang_datefunc.html).

Comment (cvogt) [inline slick/src/main/scala/slick/jdbc/SQLiteProfile.scala]: looks like this does the trick:
```
scala> new java.sql.Timestamp(javax.xml.bind.DatatypeConverter.parseDateTime("2013-10-07T08:23:19.120Z").getTime.getTime)
res3: java.sql.Timestamp = 2013-10-07 04:23:19.12
```

Comment (cvogt) [inline slick/src/main/scala/slick/jdbc/SQLiteProfile.scala]: if it doesn't parse properly you get a java.lang.IllegalArgumentException

File: slick-testkit/src/codegen/resources/dbs/sqlite.sql [hunk 5]
```diff
@@ -1,2 +1,4 @@
 CREATE TABLE "SUPPLIERS" ("SUP_ID" INTEGER NOT NULL,"SUP_NAME" VARCHAR(254) NOT NULL,"STREET" VARCHAR(254) NOT NULL,"CITY" VARCHAR(254) NOT NULL,"STATE" VARCHAR(254) NOT NULL,"ZIP" VARCHAR(254) NOT NULL);
 CREATE TABLE "COFFEES" ("COF_NAME" VARCHAR(254) NOT NULL,"SUP_ID" INTEGER NOT NULL,"PRICE" DOUBLE NOT NULL,"SALES" INTEGER NOT NULL,"TOTAL" INTEGER NOT NULL,constraint "SUP_FK" foreign key("SUP_ID") references "SUPPLIERS"("SUP_ID") on update NO ACTION on delete NO ACTION);
+CREATE TABLE "TIMESTAMP"("TS" TIMESTAMP NOT NULL DEFAULT "2016-07-29 09:14:56" , "TS2" TIMESTAMP NOT NULL DEFAULT "2016-01-01" , "TS3" TIMESTAMP NOT NULL DEFAULT 1234567890 , "TS4" TIMESTAMP NOT NULL DEFAULT "21:00:00" , "TS5" TIMESTAMP NOT NULL DEFAULT "2013-10-07T08:23:19.120Z");
+INSERT INTO "TIMESTAMP"("TS","TS2","TS3","TS4") VALUES (CURRENT_TIMESTAMP,909090,"2011-11-11","08:00:00");
```
Comment (cvogt) [inline slick-testkit/src/codegen/resources/dbs/sqlite.sql]: are these all possible allows cases for sqlite? the document also specifies "now" for example.

File: slick-testkit/src/test/scala/slick/test/codegen/GeneratedCodeTest.scala [hunk 6]
```diff
@@ -85,9 +85,12 @@ object GeneratedCodeTest {
     import CG3._
     import profile.api._
     val s = Supplier(49, "Superior Coffee", "1 Party Place", "Mendocino", "CA", "95460")
+    val ts = java.sql.Timestamp.valueOf("2016-08-30 09:14:56.909")
     DBIO.seq(
       schema.create,
       Suppliers += s,
+      Timestamp += TimestampRow( ts ),
+      Timestamp.result.head.map{ r => assertEquals(ts, r.ts) },
```
Comment (cvogt) [inline slick-testkit/src/test/scala/slick/test/codegen/GeneratedCodeTest.scala]: I guess one thing that would be nice here if we don't provide any values, let all values be generated from teh default but then read them back out and check they are correct.

File: slick/src/main/scala/slick/jdbc/SQLiteProfile.scala [hunk 7]
```diff
@@ -251,6 +263,9 @@ trait SQLiteProfile extends JdbcProfile {
       override def valueToSQLLiteral(value: Time) = value.getTime.toString
     }
     class TimestampJdbcType extends super.TimestampJdbcType {
+      override def getValue(r: java.sql.ResultSet, idx: Int) = try r.getTimestamp(idx) catch {
+        case e: java.sql.SQLException => java.sql.Timestamp.valueOf(  r.getString(idx) )
```
Comment (cvogt) [inline slick/src/main/scala/slick/jdbc/SQLiteProfile.scala]: what's the SQLException we could get here?

File: slick/src/main/scala/slick/jdbc/SQLiteProfile.scala [hunk 8]
```diff
@@ -98,14 +98,26 @@ trait SQLiteProfile extends JdbcProfile {
...
+      override def default = meta.columnDef.map((_,tpe)).collect{ 
         case ("null",_)  => Some(None) // 3.7.15-M1
+        case (v , "java.sql.Timestamp") => v match{
+          case TimestampPattern(ts , ts_unix) => Some(Some( 
+            if( ts_unix != null )
+              new java.sql.Timestamp(ts_unix.toLong)
+            else
+              try java.sql.Timestamp.valueOf(ts.replace("\"","")) catch{
+                case e: java.lang.IllegalArgumentException => 
+                    new java.sql.Timestamp(javax.xml.bind.DatatypeConverter.parseDateTime(ts).getTime.getTime)
```
Comment (lastland) [inline slick/src/main/scala/slick/jdbc/SQLiteProfile.scala]: It seems that `javax.xml.bind.DatatypeConverter.parseDateTime` is not able to parse `2013-10-07 04:23:19.120-04:00`.

**Comment by trevorsibanda** [comment]
@cvogt can you take a look at this.

`javax.xml.bind.DatatypeConverter.parseDateTime` is not able to parse `2013-10-07 04:23:19.120-04:00` but can parse `2013-10-07T04:23:19.120-04:00`

**Comment by trevorsibanda** [comment]
@cvogt @lastland - I am merging this unless there is anything you find amiss

**Comment by lastland** [comment]
I don't remember much of the previous discussion, but if all problems I have raised are fixed and a corresponding test case for each problem is added when possible, please do merge it.

**Pull request merged**